docs/faq/DesignDecisions.md

```diff
@@ -0,0 +1,75 @@
+
+Serialization enables the brower to store all actions that have been dispatched, as well as the previous store states, with much less memory. Rewinding and 'hot reloading' the store is central to the Redux developer experience and the function of Redux DevTools. This also enables deserialized actions to be stored on the server and re-serialized in the brower in the case of server-side rendering with Redux.
+
+<a id="why-currying"></a>
+### Why does the middleware signature use currying?
+The [curried function signature](https://github.com/reactjs/redux/issues/1744) of declaring middleware is [deemed unnecessary](https://github.com/reactjs/redux/pull/784) by some, because both store and next are available when the applyMiddleware function is executed. This issue has been determined to not be [worth introducing breaking changes](https://github.com/reactjs/redux/issues/1744).
+
+<a id="closure-dispatch"></a>
+### Why does applyMiddleware use a closure for dispatch?
+applyMiddleware takes the existing dispatch from the store and closes over it to create the initial chain of middlewares that have been invoked with an object that exposes the getState and dispatch functions, which enables middlewares that [rely on dispatch during initialization](https://github.com/reactjs/redux/pull/1592) to run. 
+
+<a id="combineReducers-limitations"></a>
+### Can you please change combineReducers to support nested state trees?
+No, but there are some limits to combineReducers that are worth knowing.
+- combineReducers receives an object where the values are all reducer names. Additional nesting is not possible, which limits the state shape. The following code is not possible:
+```
+const rootReducer = combineReducers({
+    a : reducerA,
+    b : {
+        b1 : reducerB1,
+        b2 : reducerB2
+    }
+});
+```
+- reducers within combineReducers are only passed the part of the state that they modify, and not the top state.
+
+The default utility combineReducers is only one way to build a complex reducer. Consider using libraries like [combineSectionReducers](https://github.com/ryo33/combine-section-reducers) or [reduceReducers](https://github.com/acdlite/reduce-reducers) if you want your reducers to have a nested tree structure.
+
+<a id="no-asynch-in-mapDispatchToProps"></a>
+### Why doesn't mapDispatchToProps allow use of return values from getState() or mapStateToProps()?
+In general, connect generates a props object out of a closure that is injected with both the current state and dispatch. Asynchronous logic does not belong in the mapStateToProps and mapDispatchToProps functions at all. They should be only pure functions which transform the state to props and bind action creators to dispatch. 
```

Comment:
Not sure I understand/follow the statements in these first two paragraphs.

I don't think we've really ever had anyone ask about calling `getState` inside of a `mapStateToProps` function.  The typical request/use case has been wanting to use either the entire state or the return value of `mapState` inside of `mapDispatch`, so that when functions are declared inside of `mapDispatch`, they can close over the latest returned values from the store.  We don't allow that approach in `mapDispatch` because it would mean also calling `mapDispatch` every time the store is updated, and also cause re-creation of functions every time, thus adding a lot of perf overhead.  

As the third paragraph says, if you _really_ want to do that, you can provide a `mergeProps` function and implement this behavior yourself, but we heavily discourage that approach.  We would much rather encourage use of component methods that pass props to bound action creators, like `this.props.toggleTodo(this.props.todoId)`.  (And, as my own personal preference, you shouldn't even write an actual `mapDispatch` function at all - instead, write separate action creator functions, and use the object shorthand form of `mapDispatch` to bind them automatically.)

Would appreciate a rework of this answer.